crates/prover-services/src/parallel/new.rs

```diff
@@ -0,0 +1,329 @@
+use std::ops::DerefMut;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use borsh::BorshDeserialize;
+use futures::future;
+use parking_lot::Mutex;
+use sov_db::ledger_db::LedgerDB;
+use sov_rollup_interface::da::{DaData, DaSpec};
+use sov_rollup_interface::services::da::DaService;
+use sov_rollup_interface::stf::StateTransitionFunction;
+use sov_rollup_interface::zk::{Proof, ZkvmHost};
+use sov_stf_runner::{
+    ProofProcessingStatus, ProverService, ProverServiceError, WitnessSubmissionStatus,
+};
+use tokio::sync::oneshot;
+
+use crate::ProofGenMode;
+
+pub(crate) type Input = Vec<u8>;
+pub(crate) type Assumptions = Vec<Vec<u8>>;
+pub(crate) type ProofData = (Input, Assumptions);
+
+/// Prover service that generates proofs in parallel.
+pub struct ParallelProverService<Da, Vm, Stf>
+where
+    Da: DaService,
+    Vm: ZkvmHost,
+    Stf: StateTransitionFunction<Vm::Guest, Da::Spec> + Send + Sync,
+    Stf::PreState: Clone + Send + Sync,
+{
+    thread_pool: rayon::ThreadPool,
+
+    proof_mode: Arc<Mutex<ProofGenMode<Da, Vm, Stf>>>,
+
+    da_service: Da,
+    vm: Vm,
+    zk_storage: Stf::PreState,
+    ledger_db: LedgerDB,
+
+    current_da_hash: Option<<Da::Spec as DaSpec>::SlotHash>,
+    proof_queue: Vec<ProofData>,
+}
+
+impl<Da, Vm, Stf> ParallelProverService<Da, Vm, Stf>
+where
+    Da: DaService,
+    Vm: ZkvmHost + 'static,
+    Stf: StateTransitionFunction<Vm::Guest, Da::Spec> + Send + Sync + 'static,
+    Stf::PreState: Clone + Send + Sync,
+{
+    /// Creates a new prover.
+    pub fn new(
+        da_service: Da,
+        vm: Vm,
+        proof_mode: ProofGenMode<Da, Vm, Stf>,
+        zk_storage: Stf::PreState,
+        thread_pool_size: usize,
+        ledger_db: LedgerDB,
+    ) -> anyhow::Result<Self> {
+        match proof_mode {
+            ProofGenMode::Skip => {
+                tracing::info!("Prover is configured to skip proving");
+            }
+            ProofGenMode::Simulate(_) => {
+                tracing::info!("Prover is configured to simulate proving");
+            }
+            ProofGenMode::Execute => {
+                tracing::info!("Prover is configured to execute proving");
+            }
+            ProofGenMode::Prove => {
+                tracing::info!("Prover is configured to prove");
+            }
+        };
+
+        let thread_pool = rayon::ThreadPoolBuilder::new()
+            .num_threads(thread_pool_size)
```

Comment:
If we have `thread_pool_size` as option, don't set this value if `None. 

>> If num_threads is 0, or you do not call this function, then the Rayon runtime will select the number of threads automatically. At present, this is based on the RAYON_NUM_THREADS environment variable (if set), or the number of logical CPUs (otherwise).

Reply:
asserted that thread_pool_size > 0